File: dart/lib/src/twitter/get_handle_and_registry_key.dart
import 'package:solana/solana.dart' hide RpcClient;

import '../constants/addresses.dart';
import '../rpc/rpc_client.dart';
import '../utils/get_domain_key_sync.dart';
import 'reverse_twitter_registry_state.dart';

/// Get Twitter handle and registry key from a verified public key
///
/// This function mirrors js/src/twitter/getHandleAndRegistryKey.ts
///
/// [connection] - The RPC client for Solana blockchain communication
/// [verifiedPubkey] - The verified public key to look up
///
/// Returns a tuple of handle and registry key or throws if not found
Future<(String, Ed25519HDPublicKey)> getHandleAndRegistryKey(
  RpcClient connection,
  Ed25519HDPublicKey verifiedPubkey,
) async {
  // Hash the verified public key (same as JS implementation)
  final hashedVerifiedPubkey = getHashedNameSync(verifiedPubkey.toBase58());

  // Get the reverse registry key using the same derivation as JS
  final reverseRegistryKeyString = await getNameAccountKeySync(
    hashedVerifiedPubkey,
    nameClass: twitterVerificationAuthority,
    nameParent: twitterRootParentRegistryAddress,
  );

  final reverseRegistryKey =
      Ed25519HDPublicKey.fromBase58(reverseRegistryKeyString);

  // Retrieve the reverse registry state (same as JS implementation)
  final reverseRegistryState = await ReverseTwitterRegistryState.retrieve(
    connection,
    reverseRegistryKey,
  );

  return (
    reverseRegistryState.twitterHandle,
    reverseRegistryState.twitterRegistryKeyPubkey,
  );
}

File: js/src/twitter/getHandleAndRegistryKey.ts
import { Connection, PublicKey } from "@solana/web3.js";
import {
  TWITTER_VERIFICATION_AUTHORITY,
  TWITTER_ROOT_PARENT_REGISTRY_KEY,
} from "../constants";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";
import { ReverseTwitterRegistryState } from "./ReverseTwitterRegistryState";

/**
 * Retrieves a verified public key's Twitter handle and user-facing registry key.
 *
 * @param connection Solana RPC connection used to retrieve the reverse registry.
 * @param verifiedPubkey Verified public key to look up.
 * @returns The Twitter handle and its user-facing registry address.
 *
 * @example
 * ```ts
 * const [handle, registry] = await getHandleAndRegistryKey(connection, owner);
 * ```
 */
export async function getHandleAndRegistryKey(
  connection: Connection,
  verifiedPubkey: PublicKey,
): Promise<[string, PublicKey]> {
  const hashedVerifiedPubkey = getHashedNameSync(verifiedPubkey.toString());
  const reverseRegistryKey = getNameAccountKeySync(
    hashedVerifiedPubkey,
    TWITTER_VERIFICATION_AUTHORITY,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  let reverseRegistryState = await ReverseTwitterRegistryState.retrieve(
    connection,
    reverseRegistryKey,
  );
  return [
    reverseRegistryState.twitterHandle,
    new PublicKey(reverseRegistryState.twitterRegistryKey),
  ];
}
